plugins/modules/mso_schema_site_vrf_region_cidr.py

```diff
@@ -178,48 +181,93 @@ def main():
+        mso.fail_json(msg="Provided vrf '{0}' does not exist. Existing vrfs: {1}".format(vrf, ', '.join(template_vrfs)))
+    else:
+        # Update vrf index at template level
+        template_vrf_idx = template_vrfs.index(vrf)
+
+    # If vrf not at site level but exists at template level
     if vrf_ref not in vrfs:
-        mso.fail_json(msg="Provided vrf '{0}' does not exist. Existing vrfs: {1}".format(vrf, ', '.join(vrfs)))
-    vrf_idx = vrfs.index(vrf_ref)
-
-    # Get Region
-    regions = [r.get('name') for r in schema_obj.get('sites')[site_idx]['vrfs'][vrf_idx]['regions']]
-    if region not in regions:
-        mso.fail_json(msg="Provided region '{0}' does not exist. Existing regions: {1}".format(region, ', '.join(regions)))
-    region_idx = regions.index(region)
-
-    # Get CIDR
-    cidrs = [c.get('ip') for c in schema_obj.get('sites')[site_idx]['vrfs'][vrf_idx]['regions'][region_idx]['cidrs']]
-    if cidr is not None and cidr in cidrs:
-        cidr_idx = cidrs.index(cidr)
-        # FIXME: Changes based on index are DANGEROUS
-        cidr_path = '/sites/{0}/vrfs/{1}/regions/{2}/cidrs/{3}'.format(site_template, vrf, region, cidr_idx)
-        mso.existing = schema_obj.get('sites')[site_idx]['vrfs'][vrf_idx]['regions'][region_idx]['cidrs'][cidr_idx]
+        op_path = '/sites/{0}/vrfs/-'.format(site_template)
+        payload.update(
+            vrfRef=dict(
+                schemaId=schema_id,
+                templateName=template,
+                vrfName=vrf,
+            ),
+            regions=[dict(
+                name=region,
+                cidrs=[new_cidr]
+            )]
+        )
+    else:
+        # Update vrf index at site level
+        vrf_idx = vrfs.index(vrf_ref)
+
+        # Get Region
+        regions = [r.get('name') for r in schema_obj.get('sites')[site_idx]['vrfs'][vrf_idx]['regions']]
+        if region not in regions:
+            op_path = '/sites/{0}/vrfs/{1}/regions/-'.format(site_template, vrf)
+            payload.update(
+                name=region,
+                cidrs=[new_cidr]
+            )
+        else:
+            region_idx = regions.index(region)
+
+            # Get CIDR
+            cidrs = [c.get('ip') for c in schema_obj.get('sites')[site_idx]['vrfs'][vrf_idx]['regions'][region_idx]['cidrs']]
+            if cidr is not None:
+                if cidr in cidrs:
+                    cidr_idx = cidrs.index(cidr)
+                    # FIXME: Changes based on index are DANGEROUS
+                    cidr_path = '/sites/{0}/vrfs/{1}/regions/{2}/cidrs/{3}'.format(site_template, vrf, region, cidr_idx)
+                    mso.existing = schema_obj.get('sites')[site_idx]['vrfs'][vrf_idx]['regions'][region_idx]['cidrs'][cidr_idx]
+                op_path = '/sites/{0}/vrfs/{1}/regions/{2}/cidrs/-'.format(site_template, vrf, region)
+                payload = new_cidr
 
     if state == 'query':
-        if cidr is None:
+        if region not in regions:
```

Comment:
what's error msg if vrf doesn't exist at site level but exist at template level?

Reply:
No error message, the VRF will be automatically created at VRF level see code from line 223 to 235.

Reply:
In my understanding, we do create region automatically while it doesn't exist in site level. But when we do query, we give an error msg when there is no site-level region. 
should we do the same in vrf?